### PR TITLE
chore: added keyring examples (kms_rsa, multi, discovery)

### PR DESCRIPTION
**Pull request by RitvikKapila**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: examples/src/keyrings/aws_kms_discovery_keyring_example.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+The AWS Encryption SDK provides a standard AWS KMS discovery keyring and a discovery keyring
+for AWS KMS multi-Region keys. For information about using multi-Region keys with the
+AWS Encryption SDK, see
+https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/configure.html#config-mrks
+
+Because it doesn't specify any wrapping keys, a discovery keyring can't encrypt data.
+If you use a discovery keyring to encrypt data, alone or in a multi-keyring, the encrypt
+operation fails. The exception is the AWS Encryption SDK for C, where the encrypt operation
+ignores a standard discovery keyring, but fails if you specify a multi-Region discovery
+keyring, alone or in a multi-keyring.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_keyring_example.py:17]:
```suggestion
operation fails. 
```

This isn't super relevant for Python, let's leave that in the docs

File: examples/src/keyrings/aws_kms_discovery_keyring_example.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+permission on the AWS KMS key.
+
+This example creates a KMS Keyring and then encrypts a custom input EXAMPLE_DATA
+with an encryption context. This encrypted ciphertext is then decrypted using the Discovery keyring.
+This example also includes some sanity checks for demonstration:
+1. Ciphertext and plaintext data are not the same
+2. Encryption context is correct in the decrypted message header
+3. Decrypted plaintext value matches EXAMPLE_DATA
+4. Decryption is only possible if the Discovery Keyring contains the correct AWS Account ID's to
+which the KMS key used for encryption belongs
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_keyring_example.py:31]:
```suggestion
4. Decryption is only possible if the Discovery Keyring contains the correct AWS Account ID's to
   which the KMS key used for encryption belongs
```

File: examples/src/keyrings/aws_kms_discovery_keyring_example.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    assert ciphertext != EXAMPLE_DATA, \
+        "Ciphertext and plaintext data are the same. Invalid encryption"
+
+    # 7. Now create a Discovery keyring to use for decryption. We'll add a discovery filter
+    # so that we limit the set of ciphertexts we are willing to decrypt to only ones
+    # created by KMS keys in our account and partition.
+    discovery_keyring_input: CreateAwsKmsDiscoveryKeyringInput = CreateAwsKmsDiscoveryKeyringInput(
+        kms_client=kms_client,
+        discovery_filter=DiscoveryFilter(
+            account_ids=["658956600833"],
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_keyring_example.py:135]: ! Let's not hardcode our account ID directly in the test.
Can you think of a better way to provide the account ID?

File: examples/src/keyrings/aws_kms_discovery_keyring_example.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    kms_keyring_input: CreateAwsKmsKeyringInput = CreateAwsKmsKeyringInput(
+        kms_key_id=kms_key_id,
+        kms_client=kms_client
+    )
+
+    encrypt_kms_keyring: IKeyring = mat_prov.create_aws_kms_keyring(
+        input=kms_keyring_input
+    )
+
+    # 5. Encrypt the data for the encryptionContext
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_keyring_example.py:117]:
```suggestion
    # 5. Encrypt the data with the encryptionContext
```

I'll stop leaving this comment, change it on the others

File: examples/src/keyrings/aws_kms_discovery_keyring_example.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            "Encryption context does not match expected values"
+
+    # 10. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes == EXAMPLE_DATA
+
+    # 11. Demonstrate that if a discovery keyring (Bob's) doesn't have the correct AWS Account ID's,
+    # the decrypt will fail with an error message
+    # Note that the Account ID '888888888888' used here is different than the one used
+    # during encryption '658956600833'
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_keyring_example.py:175]:
```suggestion
    # Note that this assumes Account ID used here ('888888888888') is different than the one used
    # during encryption
```

File: examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+The AWS Encryption SDK provides a standard AWS KMS discovery keyring and a discovery keyring
+for AWS KMS multi-Region keys. For information about using multi-Region keys with the
+AWS Encryption SDK, see
+https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/configure.html#config-mrks
+
+Because it doesn't specify any wrapping keys, a discovery keyring can't encrypt data.
+If you use a discovery keyring to encrypt data, alone or in a multi-keyring, the encrypt
+operation fails. The exception is the AWS Encryption SDK for C, where the encrypt operation
+ignores a standard discovery keyring, but fails if you specify a multi-Region discovery
+keyring, alone or in a multi-keyring.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py:16]:
```suggestion
operation fails.
```

File: examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        "Ciphertext and plaintext data are the same. Invalid encryption"
+
+    # 7. Now create a Discovery Multi keyring to use for decryption. We'll add a discovery filter
+    # so that we limit the set of ciphertexts we are willing to decrypt to only ones
+    # created by KMS keys in our account and partition.
+    discovery_multi_keyring_input: CreateAwsKmsDiscoveryMultiKeyringInput = \
+        CreateAwsKmsDiscoveryMultiKeyringInput(
+            regions=["us-west-2", "us-east-1"],
+            discovery_filter=DiscoveryFilter(
+                account_ids=["658956600833"],
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py:132]: Ditto, refactor account ID

File: examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert ciphertext != EXAMPLE_DATA, \
+        "Ciphertext and plaintext data are the same. Invalid encryption"
+
+    # 7. Now create a Discovery Multi keyring to use for decryption. We'll add a discovery filter
+    # so that we limit the set of ciphertexts we are willing to decrypt to only ones
+    # created by KMS keys in our account and partition.
+    discovery_multi_keyring_input: CreateAwsKmsDiscoveryMultiKeyringInput = \
+        CreateAwsKmsDiscoveryMultiKeyringInput(
+            regions=["us-west-2", "us-east-1"],
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py:130]: Let's refactor `regions` too and provide it the same way we provide the account ID

File: examples/src/keyrings/aws_kms_multi_keyring_example.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+When decrypting, the AWS Encryption SDK uses the keyrings to try to decrypt one of the encrypted
+data keys. The keyrings are called in the order that they are specified in the multi-keyring.
+Processing stops as soon as any key in any keyring can decrypt an encrypted data key.
+
+This example creates a Multi Keyring and then encrypts a custom input EXAMPLE_DATA
+with an encryption context. This example also includes some sanity checks for demonstration:
+1. Ciphertext and plaintext data are not the same
+2. Decryption of ciphertext is possible using the multi_keyring,
+and every one of the keyrings from the multi_keyring separately
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_multi_keyring_example.py:29]:
```suggestion
   and every one of the keyrings from the multi_keyring separately
```

File: examples/src/keyrings/aws_kms_multi_keyring_example.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    # 7d. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes_default_region_kms_keyring == EXAMPLE_DATA
+
+    # 8. Demonstrate that you can also successfully decrypt data using a KMS keyring with just the
+    # `second_region_kms_key_id` directly.
+
+    # 8a. Create a boto3 client for KMS for the second region.
+    second_region_kms_client = boto3.client('kms', region_name="eu-central-1")
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_multi_keyring_example.py:173]: refactor the region out

File: examples/src/keyrings/aws_kms_rsa_keyring_example.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+    # 4. Create a KMS RSA keyring
+    mat_prov: AwsCryptographicMaterialProviders = AwsCryptographicMaterialProviders(
+        config=MaterialProvidersConfig()
+    )
+
+    keyring_input: CreateAwsKmsRsaKeyringInput = CreateAwsKmsRsaKeyringInput(
+        public_key=public_key,
+        kms_key_id=kms_rsa_key_id,
+        encryption_algorithm="RSAES_OAEP_SHA_256",
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_rsa_keyring_example.py:87]: Could you link to the boto3 docs to source this value?

**Review by lucasmcdonald3 — COMMENTED** [review]
Let's do smaller PRs next time. 5 is a little much.
If some of these were in their own PRs, they could be merged in sooner, since some of this examples are in better shape than others. Since they're all together, they're all blocked from merge by the example in worst shape.
Keep it at 2-3 examples per PR.

File: examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    Usage: get_account_id_from_kms_key_id(kms_key_id)
+    :param kms_key_id: KMS Key identifier for the KMS key you want to use
+    :type kms_key_id: string
+    :return: AWS Account ID
+    :rtype: string
+    """
+    return kms_key_id.split(":")[4]
+
+
+def get_aws_region_from_kms_key_id(kms_key_id: str) -> str:
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_discovery_multi_keyring_example.py:69]: Again, I can see why you would choose this approach, but let's just provide the region as a separate input to the functions.
Parsing the key ARN for metadata isn't a best practice, and we don't want to model this behavior for customers.

File: examples/src/keyrings/aws_kms_rsa_keyring_example.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        "that can help you": "be confident that",
+        "the data you are handling": "is what you think it is",
+    }
+
+    # 4. Create a KMS RSA keyring
+    mat_prov: AwsCryptographicMaterialProviders = AwsCryptographicMaterialProviders(
+        config=MaterialProvidersConfig()
+    )
+
+    # # Create the AWS KMS RSA keyring input
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_rsa_keyring_example.py:84]:
```suggestion
    # Create the AWS KMS RSA keyring input
```

File: examples/src/keyrings/aws_kms_multi_keyring_example.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+    # 6b. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes_multi_keyring == EXAMPLE_DATA
+
+    # Because you used a multi_keyring on Encrypt, you can use either of the two
+    # kms keyrings individually to decrypt the data.
+
+    # 7. Demonstrate that you can successfully decrypt data using a KMS keyring with just the
+    # `default_region_kms_key_id` directly.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_multi_keyring_example.py:157]:
```suggestion
    # 7. Demonstrate that you can successfully decrypt data using a KMS keyring with just the
    # `default_region_kms_key_id` directly.
    # (This is an example for demonstration; you do not need to do this in your own code.)
```

File: examples/src/keyrings/aws_kms_multi_keyring_example.py [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+        source=ciphertext,
+        keyring=default_region_kms_keyring
+    )
+
+    # 7d. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes_default_region_kms_keyring == EXAMPLE_DATA
+
+    # 8. Demonstrate that you can also successfully decrypt data using a KMS keyring with just the
+    # `second_region_kms_key_id` directly.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/aws_kms_multi_keyring_example.py:184]:
```suggestion
    # 8. Demonstrate that you can also successfully decrypt data using a KMS keyring with just the
    # `second_region_kms_key_id` directly.
    # (This is an example for demonstration; you do not need to do this in your own code.)
```

File: examples/src/keyrings/multi_keyring_example.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+    # 10b. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes_multi_keyring == EXAMPLE_DATA
+
+    # Because you used a multi_keyring on Encrypt, you can use either the
+    # `kms_keyring` or `raw_aes_keyring` individually to decrypt the data.
+
+    # 11. Demonstrate that you can successfully decrypt data using just the `kms_keyring`
+    # directly.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/multi_keyring_example.py:184]:
```suggestion
    # 11. Demonstrate that you can successfully decrypt data using just the `kms_keyring`
    # directly.
    # (This is an example for demonstration; you do not need to do this in your own code.)
```

File: examples/src/keyrings/multi_keyring_example.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+        source=ciphertext,
+        keyring=kms_keyring
+    )
+
+    # 11b. Demonstrate that the decrypted plaintext is identical to the original plaintext.
+    # (This is an example for demonstration; you do not need to do this in your own code.)
+    assert plaintext_bytes_kms_keyring == EXAMPLE_DATA
+
+    # 12. Demonstrate that you can also successfully decrypt data using the `raw_aes_keyring`
+    # directly.
```
Comment (lucasmcdonald3) [inline examples/src/keyrings/multi_keyring_example.py:197]:
```suggestion
    # 12. Demonstrate that you can also successfully decrypt data using the `raw_aes_keyring`
    # directly.
    # (This is an example for demonstration; you do not need to do this in your own code.)
```

File: examples/src/keyrings/aws_kms_rsa_keyring_example.py (hunk 13 above)
Comment (RitvikKapila) [reply examples/src/keyrings/aws_kms_rsa_keyring_example.py:84]: That's a good catch!

**Review by lucasmcdonald3 — APPROVED** [review]

**Pull request merged**